haskell-src/Concordium/MerkleProofs.hs

```diff
@@ -0,0 +1,566 @@
+--  of the byte string.
+--
+--  Note that a 'MerkleProof' can contain multiple 'RawData' chunks in sequence, which is equivalent
+--  to a single 'RawData' chunk with the 'BS.ByteString's concatenated. It can also contain
+--  'RawData' chunks with zero-length 'BS.ByteString's, which can always be omitted to produce an
+--  equivalent 'MerkleProof'. For storage and transmission, compact representations should be
+--  preferred, but functions operating on 'MerkleProof's should tolerate all representations.
+--
+--  This representation is broadly agnostic to what the proof represents, which will depend on
+--  the structure of the original Merkle tree. Moreover, a 'MerkleProof' allows multiple paths
+--  to be revealed in the same proof.
+type MerkleProof = [MerkleBranch]
+
+-- | A branch in a Merkle proof.
+data MerkleBranch
+    = -- | Raw bytes that are concatenated as-is when computing the root hash.
+      RawData !BS.ByteString
+    | -- | A subproof that is reduced to a hash before concatenating when computing the root hash.
+      SubProof !MerkleProof
+    deriving (Show)
+
+-- | Determine if a Merkle proof is empty (it contains no bytes).
+isEmpty :: MerkleProof -> Bool
+isEmpty [] = True
+isEmpty (RawData bytes : rest) = BS.null bytes && isEmpty rest
+isEmpty _ = False
+
+-- | Compute the root hash represented by a Merkle proof.
+toRootHash :: MerkleProof -> SHA256.Hash
+toRootHash l = SHA256.hashLazy $ Builder.toLazyByteString $ mconcat $ merkleBuilder <$> l
+
+-- | A helper function for constructing a 'Builder.Builder' from a 'MerkleBranch'.
+-- This is used for managing branches for the purposes of computing root hashes.
+merkleBuilder :: MerkleBranch -> Builder.Builder
+merkleBuilder (RawData bytes) = Builder.byteString bytes
+merkleBuilder (SubProof sp) = Builder.shortByteString (SHA256.hashToShortByteString $ toRootHash sp)
+
+-- * Parsing Merkle proofs
+
+-- | How a size is encoded in a Merkle proof.
+-- (All sizes are unsigned, since a negative size is meaningless.)
+data SizeEncoding
+    = -- | 16-bit big-endian
+      SizeU16BE
+    | -- | 32-bit big-endian
+      SizeU32BE
+    | -- | 64-bit big-endian
+      SizeU64BE
+    deriving (Eq, Show)
+
+-- | A schematic type that describes how some data item is represented in a Merkle proof.
+-- We do not fully parse data, but just tokenise it to fixed or variable length byte strings.
+data MerkleData
+    = -- | A byte string of the given fixed length.
+      FixedLengthBytes Word64
+    | -- | A variable length string of bytes prefixed by its length, encoded based on the size
+      -- encoding.
+      VariableLengthBytes SizeEncoding
+    deriving (Show)
+
+-- | Tag type used to label branches in the parse tree of a Merkle proof.
+type Tag = String
+
+-- | Identifier type used to identify non-terminals in a Merkle schema.
+type NonTerminal = Int
+
+-- | The body of a rewrite rule from a 'MerkleSchema'.
```

Comment:
It's unclear waht "rewrite rule" refers to here. What is the system?

Or in other words, this comment was not illuminating without more context.